overlord/snapstate/handlers.go

```diff
@@ -3467,25 +3535,54 @@ func InjectAutoConnect(mainTask *state.Task, snapsup *SnapSetup) {
 	mainTask.Logf("added auto-connect task")
 }
 
-// GetSnapDirOptions returns *dirs.SnapDirOptions configured according to the
-// enabled experimental features. The state must be locked by the caller.
-var GetSnapDirOptions = getSnapDirOptions
+type dirMigrationOptions struct {
+	// UseHidden states whether the user has requested that the hidden data dir be used
+	UseHidden bool
+
+	// MigratedToHidden states whether the data has been migrated to the hidden dir
+	MigratedToHidden bool
+}
 
-func getSnapDirOptions(state *state.State) (*dirs.SnapDirOptions, error) {
-	tr := config.NewTransaction(state)
+// GetSnapDirOpts returns the snap dir options based on the current the migration status
+func (o *dirMigrationOptions) getSnapDirOpts() *dirs.SnapDirOptions {
+	return &dirs.SnapDirOptions{HiddenSnapDataDir: o.MigratedToHidden}
+}
 
+// GetSnapDirOpts returns the options required to get the correct snap dir.
+var GetSnapDirOpts = func(st *state.State, name string) (*dirs.SnapDirOptions, error) {
+	var snapst SnapState
+	if err := Get(st, name, &snapst); err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, err
+	}
+
+	hiddenOpts, err := getDirMigrationOpts(st, &snapst, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return hiddenOpts.getSnapDirOpts(), nil
+}
+
+// getDirMigrationOpts checks if the feature flag is set and if the snap data
+// has been migrated, first checking the SnapSetup (if not nil) and then
+// the SnapState. The state must be locked by the caller.
+var getDirMigrationOpts = func(st *state.State, snapst *SnapState, snapsup *SnapSetup) (*dirMigrationOptions, error) {
+	tr := config.NewTransaction(st)
 	hiddenDir, err := features.Flag(tr, features.HiddenSnapDataHomeDir)
 	if err != nil {
 		return nil, fmt.Errorf("cannot read feature flag %q: %w", features.HiddenSnapDataHomeDir, err)
 	}
 
-	return &dirs.SnapDirOptions{HiddenSnapDataDir: hiddenDir}, nil
-}
+	opts := &dirMigrationOptions{UseHidden: hiddenDir}
 
-func MockGetSnapDirOptions(f func(*state.State) (*dirs.SnapDirOptions, error)) (restore func()) {
-	old := getSnapDirOptions
-	GetSnapDirOptions = f
-	return func() {
-		GetSnapDirOptions = old
+	if snapst != nil {
```

Comment:
I think snapst should never be nil here.  Also just for completeness my suggestion was to do something like:

```
if snapst.IsInstalled() {
   opts.MigratedToHidden = snapst.MigratedHidden
} else {
  opts.MigratedToHidden = opts.UseHidden
}
```
as you said it might be overkill anyway.

Reply:
Yes, it's never called with nil (except in tests) and initially I built the struct with `MigratedToHidden: snapst.MigratedHidden` but thought it wouldn't hurt to be extra paranoid and add a nil check